app/drivers/sensor/battery/battery_nrf_vddh.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2021 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ *
+ * This is a simplified version of battery_voltage_divider.c which always reads
+ * the VDDHDIV5 channel of the &adc node and multiplies it by 5.
+ */
+
+#define DT_DRV_COMPAT zmk_battery_nrf_vddh
+
+#include <device.h>
+#include <devicetree.h>
+#include <drivers/adc.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+
+#include "battery_common.h"
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#define VDDHDIV (5)
+
+struct vddh_data {
+    const struct device *adc;
+    struct adc_channel_cfg acc;
+    struct adc_sequence as;
+    uint16_t adc_raw;
+    uint16_t millivolts;
+    uint8_t state_of_charge;
+};
+
+static int vddh_sample_fetch(const struct device *dev, enum sensor_channel chan) {
+    // Make sure selected channel is supported
+    if (chan != SENSOR_CHAN_GAUGE_VOLTAGE && chan != SENSOR_CHAN_GAUGE_STATE_OF_CHARGE &&
+        chan != SENSOR_CHAN_ALL) {
+        LOG_DBG("Selected channel is not supported: %d.", chan);
+        return -ENOTSUP;
+    }
+
+    struct vddh_data *drv_data = dev->data;
+    struct adc_sequence *as = &drv_data->as;
+
+    int rc = adc_read(drv_data->adc, as);
+    as->calibrate = false;
+
+    if (rc != 0) {
+        LOG_ERR("Failed to read ADC: %d", rc);
+        return rc;
+    }
+
+    int32_t val = drv_data->adc_raw;
+    rc = adc_raw_to_millivolts(adc_ref_internal(drv_data->adc), drv_data->acc.gain, as->resolution,
+                               &val);
+    if (rc != 0) {
+        LOG_ERR("Failed to convert raw ADC to mV: %d", rc);
+        return rc;
+    }
+
+    drv_data->millivolts = val * VDDHDIV;
+    drv_data->state_of_charge = lithium_ion_mv_to_pct(drv_data->millivolts);
+
+    LOG_DBG("ADC raw %d ~ %d mV => %d%%", drv_data->adc_raw, drv_data->millivolts,
+            drv_data->state_of_charge);
+
+    return rc;
+}
+
+static int vddh_channel_get(const struct device *dev, enum sensor_channel chan,
+                            struct sensor_value *val) {
+    struct vddh_data const *drv_data = dev->data;
+
+    switch (chan) {
+    case SENSOR_CHAN_GAUGE_VOLTAGE:
+        val->val1 = drv_data->millivolts / 1000;
+        val->val2 = (drv_data->millivolts % 1000) * 1000U;
+        break;
+
+    case SENSOR_CHAN_GAUGE_STATE_OF_CHARGE:
+        val->val1 = drv_data->state_of_charge;
+        val->val2 = 0;
+        break;
+
+    default:
+        return -ENOTSUP;
+    }
+
+    return 0;
+}
```

Comment:
It seems to me that this function is basically identical to the voltage divider driver. Can we bring this function to `battery_common.c`?

Reply:
The only reason I hadn't done that already was that the driver data structs are different. I could pull out the code that populates val given a state of charge or a millivolts value pretty easily though.

Reply:
Switch both battery drivers to using a common data struct and pulled this out into a common function.